File: Juego/Invasiones/fuente/Recursos/Texto.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Xml;
using System.IO;
using Invasiones.Debug;

namespace Invasiones.Recursos
{
    /// <summary>
    /// Clase utilizada para cargar todos los textos, desde el archivo xml.
    /// Cachea todos los strings.
    /// </summary>
    public class Texto
    {
        #region Metodos Static
        /// <summary>
        /// Contiene todos los strings.
        /// </summary>
        private static string [] s_strings;

        /// <summary>
        ///Carga el texto básico para correr la aplicación.
        /// </summary>
        public static bool Cargar()
        {
            s_strings = new string[Res.STR_COUNT];
            string path = Utilidades.ObtenerPath(Programa.ARCHIVO_XML_TEXTOS);

            if (path == null)
            {
                Log.Instancia.Error("No se encuentra el archivo " + Programa.ARCHIVO_XML_TEXTOS + ".");
                return false;
            }

            try
            {
                XmlTextReader lector = new XmlTextReader(path);

                lector.MoveToContent();
                lector.ReadStartElement();
                int i = 0;
                while (lector.Read())
                {
                    if (lector.NodeType == XmlNodeType.Element)
                    {
                        s_strings[i++] = lector.ReadString();
                    }
                }
                lector.Close();
            }
            catch (Exception e)
            {
                Log.Instancia.Error("Error al leer el archivo" + Programa.ARCHIVO_XML_TEXTOS + "." +
                      e.Message);
                return false;
            }
            return true;
        }

        ///<summary>
        ///Devuelve los strings utilizados.
        ///</summary>
        public static string[] Strings
        {
            get
            {
                if (s_strings == null)
                    
                {
                    Cargar();
                }
                return s_strings;
            }
        }
        #endregion
    }
}

File: macOS/Invasiones/Invasiones/Recursos/Texto.swift
// Recursos/Texto.swift
// Puerto de Texto.cs — carga y caché de strings localizados desde strings.xml.
// Usa XMLParser (SAX) en lugar de XmlTextReader de .NET.

import Foundation

class Texto: NSObject, XMLParserDelegate {

    // MARK: - Static storage
    private static var s_strings: [String]?

    /// Carga los strings desde strings.xml en el bundle.
    @discardableResult
    static func cargar() -> Bool {
        s_strings = Array(repeating: "", count: Res.STR_COUNT)

        guard let path = Utilidades.obtenerPath(Programa.ARCHIVO_XML_TEXTOS) else {
            Log.Instancia.error("No se encuentra el archivo \(Programa.ARCHIVO_XML_TEXTOS).")
            return false
        }

        let parser = Texto()
        let xmlParser = XMLParser(contentsOf: URL(fileURLWithPath: path))
        xmlParser?.delegate = parser
        let ok = xmlParser?.parse() ?? false

        if !ok {
            Log.Instancia.error("Error al leer el archivo \(Programa.ARCHIVO_XML_TEXTOS).")
            return false
        }

        s_strings = parser.m_stringsLeidos
        return true
    }

    /// Devuelve el array de strings (carga lazy si aún no fue cargado).
    static var Strings: [String] {
        if s_strings == nil {
            cargar()
        }
        return s_strings ?? []
    }

    // MARK: - XMLParserDelegate (instancia temporal usada solo durante el parseo)
    private var m_stringsLeidos: [String] = Array(repeating: "", count: Res.STR_COUNT)
    private var m_indice = 0
    private var m_textoActual = ""
    private var m_enElemento = false

    func parser(_ parser: XMLParser, didStartElement elementName: String,
                namespaceURI: String?, qualifiedName qName: String?,
                attributes attributeDict: [String: String] = [:]) {
        // Saltar el elemento raíz (<strings>)
        if elementName.lowercased() != "strings" {
            m_textoActual = ""
            m_enElemento = true
        }
    }

    func parser(_ parser: XMLParser, foundCharacters string: String) {
        if m_enElemento {
            m_textoActual += string
        }
    }

    func parser(_ parser: XMLParser, didEndElement elementName: String,
                namespaceURI: String?, qualifiedName qName: String?) {
        if m_enElemento && m_indice < Res.STR_COUNT {
            let trimmed = m_textoActual.trimmingCharacters(in: .whitespacesAndNewlines)
            // strings.xml usa la secuencia literal "\n"; convertirla a salto de línea real.
            m_stringsLeidos[m_indice] = trimmed.replacingOccurrences(of: "\\n", with: "\n")
            m_indice += 1
            m_enElemento = false
        }
    }
}
